lib/Onyx.js

```diff
@@ -981,32 +1008,27 @@ function applyMerge(key, data) {
  * Onyx.merge(ONYXKEYS.POLICY, {name: 'My Workspace'}); // -> {id: 1, name: 'My Workspace'}
  *
  * @param {String} key ONYXKEYS key
- * @param {(Object|Array)} value Object or Array value to merge
+ * @param {(Object|Array)} changes Object or Array value to merge
  * @returns {Promise}
  */
-function merge(key, value) {
-    if (mergeQueue[key]) {
-        mergeQueue[key].push(value);
-        return Promise.resolve();
-    }
-
-    mergeQueue[key] = [value];
-    return get(key)
-        .then((data) => {
+function merge(key, changes) {
+    const nextModification = () => get(key)
+        .then((existingValue) => {
             try {
-                const modifiedData = applyMerge(key, data);
+                const modifiedData = applyMerge(existingValue, changes);
 
-                // Clean up the write queue so we
-                // don't apply these changes again
-                delete mergeQueue[key];
-
-                return set(key, modifiedData);
+                // We relay changes to subscribers optimistically to avoid lag (think UI updates)
+                notify(key, modifiedData, 'merge');
```

Comment:
Am I wrong in thinking that this logic will now get called n times per concurrent key we try to update the value for?

Reply:
Should we change this behaviour?

Reply:
FYI, we have some logic in App now which expects that all of the updates in a list are applied before notifying the subscribers. I believe it works for most cases, but it's not perfect. I've thought about adding a `batchUpdate` function to Onyx, but Marc explained that we probably don't need it since calls to merge are effectively batched from the perspective of the subscriber. We also situations where there are multiple merge updates sent for the same object so I think it would be good to keep the existing behavior.

Reply:
BTW [here](https://github.com/Expensify/App/blob/36102d192643703cd83ce0f620f6639ac877a382/src/libs/Network/SequentialQueue.js#L86) is where we are expecting that all updates are applied before notifying subscribers.

I've observed that this doesn't always work as expected and I would appreciate some input on [this comment](https://github.com/Expensify/App/issues/15550#issuecomment-1513933360) if you have time.

Anyways, I think we should keep the current behavior where multiple merges in the span of a single tick would accumulate and then notify.

Reply:
This would bring us back to the original `mergeQueue`.

The only way i could think of to make `JSON_PATCH` work with this type of queuing, would be to only do a native merge (JSON_PATCH) if we have a single item in the queue, a.k.a. only one delta that doesn't need to be combined before actually sending it to Storage.

If multiple merge values exist in `mergeQueue`, we rather merge the value via `applyMerge` and use `Storage.setItem` instead of `Storage.mergeItem`

Reply:
Like @neil-marcellini 's [proposal](https://github.com/Expensify/react-native-onyx/pull/238#discussion_r1204618185), would it be good idea to wait with notifying until after the write operation is done and extract setting the cache out of the `notify` function?

This way, we could spare the `applyMerge` call for single merge items in the queue, since we just notify with what `JSON_PATCH` returns to us as the merge result.

For multiple items in the queue, we'd still need `applyMerge`